### PR TITLE
Add datamodels

### PR DESCRIPTION
**Pull request by egparedes**
## Description

Add a new eve.datamodels module to replace pydantic.Models in the short future.

Main changes:
- Add eve.datamodels module with unit tests
- Add new typing manipulation utils to eve.typingx
- Add new minor utils to eve.utils
- Very minor fixes to imports and docs in some eve modules

### CODE REVIEW

**Comment by egparedes** [comment]
Original review by @havogt in gtc repo: https://github.com/GridTools/gtc/pull/97

**Comment by egparedes** [comment]
bors try

**Comment by egparedes** [comment]
bors try

File: tests/eve_tests/unit_tests/test_datamodels.py [hunk 1]
```diff
@@ -0,0 +1,842 @@
...
+class GlobalRecursiveModel:
+    value: Optional[GlobalRecursiveModel] = None
+
+
+def test_deferred_class_type_hint():
+    # Model defined in a module global context
+    assert isinstance(GlobalRecursiveModel.__datamodel_fields__.value.type, ForwardRef)
+
+    m1 = GlobalRecursiveModel()
+    m2 = GlobalRecursiveModel(value=m1)
```
Comment (tehrengruber) [inline tests/eve_tests/unit_tests/test_datamodels.py:352]: Maybe a test for a model with a `Union` or `List` of models as a field would be useful to test the recursion of the type validator.

File: src/eve/datamodels.py [hunk 2]
```diff
@@ -0,0 +1,1290 @@
...
+
+    Examples:
+        >>> from typing import Generic, get_origin
+        >>> @datamodel
+        ... class Model(Generic[T]):
+        ...     value: T
+        >>> assert isinstance(Model[int], GenericDataModelAlias)
+        >>> assert issubclass(get_origin(Model[int]), Model)
+        >>> assert Model[int].__class__ is get_origin(Model[int])
+        >>> print(Model[int].__class__.__name__)
```
Comment (tehrengruber) [inline src/eve/datamodels.py:189]: An example to retrieve the type parameters, e.g. `T` / `int` from `Model[int]`, would be nice. In `typing._GenericAlias` this information is not exposed, but useful occasionally.

File: tests/eve_tests/unit_tests/test_datamodels.py [hunk 3]
```diff
@@ -0,0 +1,842 @@
...
+    assert isinstance(Model.__datamodel_fields__.value.metadata, types.MappingProxyType)
+    assert Model.__datamodel_fields__.value.metadata["my_metadata"] == "META"
+
+
+# Test datamodel options
+def test_frozen():
+    import attr
+
+    @datamodels.datamodel(frozen=True)
+    class FrozenModel:
```
Comment (tehrengruber) [inline tests/eve_tests/unit_tests/test_datamodels.py:668]: What is the behavior for a frozen model with a mutable model as a field, is that allowed?

File: tests/eve_tests/unit_tests/test_datamodels.py [hunk 4]
```diff
@@ -0,0 +1,842 @@
...
+def test_hash():
+    string_value = "this is a really long string to avoid string interning 1234567890 +:?.!"
+
+    # Safe mutable case: no hash
+    @datamodels.datamodel
+    class MutableModel:
+        value: Any = None
+
+    with pytest.raises(TypeError, match="unhashable type"):
+        hash(MutableModel(value=string_value))
```
Comment (tehrengruber) [inline tests/eve_tests/unit_tests/test_datamodels.py:692]: Seems useful to document the motivation behind this behavior. This effectively prevents the creating of dictionaries with mutable models as keys. This case however should be prevented as attaching information to models which are mutable might be unsafe. When the model is mutated the attached information might not be valid anymore.

**Review by tehrengruber — APPROVED** [review]
It would be nice if we could at some point have type checks for this case:
```
class SomeDecl:
  ...

class SymbolRef(eve.dataclass, Generic[T]):
  name: SymbolRef
  decl: T

class SomeModel:
  field: SymbolRef[SomeDecl]
```
Looks really good, just a few comments.

File: tests/eve_tests/unit_tests/test_datamodels.py (hunk 3 above)
Comment (egparedes) [reply tests/eve_tests/unit_tests/test_datamodels.py:668]: Yes, right now the behavior is the same as dataclass/attrs, that is, the instance becomes immutable but not its fields (just like a tuple of lists or other mutable items). In the future I plan to add automatic converters to also freeze the supported collection types (sequences, sets and mappings).

File: tests/eve_tests/unit_tests/test_datamodels.py (hunk 4 above)
Comment (egparedes) [reply tests/eve_tests/unit_tests/test_datamodels.py:692]: This is the default Python / attrs / dataclasses behavior. I will add a link for reference.

File: src/eve/datamodels.py (hunk 2 above)
Comment (egparedes) [reply src/eve/datamodels.py:189]: I added a small example of the expected results for parameters/args when dealing with generic types, but that's all. The mapping from type variables to concrete types is not exposed in `GenericAlias`, there is nothing we could here if we want to keep 100%  compatibilty .

File: tests/eve_tests/unit_tests/test_datamodels.py (hunk 1 above)
Comment (egparedes) [reply tests/eve_tests/unit_tests/test_datamodels.py:355]: Done

**Comment by egparedes** [comment]
bors try

File: tests/eve_tests/unit_tests/test_datamodels.py (hunk 1 above)
Comment (egparedes) [reply tests/eve_tests/unit_tests/test_datamodels.py:355]: By the way, `Optional[T]` is already a Union (`Union[T, NoneType]`)  :)

File: src/eve/datamodels.py (hunk 2 above)
Comment (tehrengruber) [reply src/eve/datamodels.py:195]: Correct me if I'm wrong, but `_GenericAlias` actually does expose this in `__args__`. It's just that `_GenericAlias` itself is not public. I do however see that what we expose here is on par with what is exposed indirectly in `_GenericAlias` via `get_origin`.

**Pull request merged**